### PR TITLE
Refactor device interfaces

### PR DESCRIPTION
**Pull request by v0l**
### Description
This PR refactors the device communication and operations interfaces to be more generic (remove DFU like interface)

This is a follow on from #8 

Current state of this PR includes listing serial device on windows:
```
> radio_tool.exe -l
[0483:df11]: idx=000, AnyRoad Technology Digital Radio in USB mode
[COM1]: idx=1, Generic serial radio
[COM4]: idx=2, Generic serial radio
```

This is currently listing all serial ports as options, hopefully i can find a way to add `VID:PID` filtering on top of the full serial device list.

### Dependency graph
Basic interface looks like this:
```
[RadioFactory] -> [RadioOperations] -> [RadioDevice]
```

Implementations
```
[RadioFactory]
- USBRadioFactory
- SerialRadioFactory
```
```
[RadioOperations]
- TYTRadio
- AilunceRadio
```
```
[RadioDevice]
- TYTDevice
- YModemDevice
```

### CODE REVIEW

**Comment by v0l** [comment]
@n1zzo Could you test this on a windows machine for `HD1` radio? I added some windows serial code but i dont have this radio to test, it seems like its trying to transmit data on my serial ports but i cant be sure that its working.

**Comment by n1zzo** [comment]
Hi v0l,
I'm getting this error message:
immagine
Which additional library do I need to run `radio_tool`?

**Comment by v0l** [comment]
You need to install MSVC Runtime:

Try this one: https://aka.ms/vs/17/release/vc_redist.x64.exe

**Comment by n1zzo** [comment]
Uhm, nope I still see the error on both my windows machines,
here are the MSVC versions I have installed:

immagine

**Comment by v0l** [comment]
`MSVCP140D.dll` is the C++ runtime dll so not sure why its not working for you... I was able to run it ok..  

You could also build it on your machine but you need to install `CMake` and `vcpkg`

image

**Comment by v0l** [comment]
@n1zzo Any luck? Also if you could test on any other OS you have access from. 

Or maybe you can ask other you know to test the build: https://github.com/v0l/radio_tool/releases/tag/v0.2.0-beta.1

**Comment by n1zzo** [comment]
Hi @v0l I will test tomorrow on linux and ask the community to test on WIndows.

Congratulations again for the nice work you are putting into this project!

Niccolò

**Comment by n1zzo** [comment]
Hi @v0l,
sorry for the late review. Flashing from linux leads to a segfault. The issue is in the file `fymodem.c`:
```C
 56 char __ym_getchar(int timeout_ms)
 57 {
 58     char c = 0;
 59 #ifdef _WIN32
 60     ReadFile((HANDLE)global_fd, &c, (DWORD)1, NULL, NULL);
 61 #else
 62     fread(&c, 1, 1, global_fd);
 63 #endif
 64     return c;
 65 }
```
The `fread` on line 62 is segfaulting.
I will try to fix the issue.

**Comment by n1zzo** [comment]
Found the issue, `fread` expects a FILE* pointer, you're passing a file descriptor instead.

**Comment by v0l** [comment]
ah, probably i didn't finish converting this to `FILE*` type then or got mixed up somewhere.

**Comment by n1zzo** [comment]
Yeah,
probably, let me know when the fix is out so that I can test it! :wink:

File: src/ymodem_device.cpp [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+YModemDevice::YModemDevice(const std::string &port, const std::string &filename) : port(port), filename(filename), fd(-1)
+{
+    int fdOpen = -1;
+#ifdef _WIN32
+    std::stringstream portPath;
+    portPath << "\\\\.\\" << port;
+
+    fdOpen = (int)CreateFileA(portPath.str().c_str(), GENERIC_READ | GENERIC_WRITE, NULL, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+#else
+    fdOpen = open(port.c_str(), O_RDWR | O_NOCTTY | O_SYNC);
```
Comment (v0l) [inline src/ymodem_device.cpp:44]: Should be `FILE?`

Comment (n1zzo) [reply src/ymodem_device.cpp:44]: Yeah, if you want to use `FILE *` everywhere this should become a `fopen`!

**Comment by v0l** [comment]
@n1zzo I pushed some fixes here: 0afaac56de44f1370cd941585e55deef69a9f782 

Just went back to using FD, sems to work on Macbook..

**Comment by n1zzo** [comment]
I confirm that this works also on linux!
Thanks, will point OpenRTX toolchain to the new version of `radio_tool`.

**Pull request merged**